internal/tasks/manifests.go

```diff
@@ -891,3 +893,178 @@ func (j *Janitor) setManifestAndParentsToPending(ctx context.Context, manifestDi
+	}).Setup(registerer)
+}
+
+func (j *Janitor) processTrivySecurityInfo(tx *gorp.Transaction, securityInfo keppel.TrivySecurityInfo, labels prometheus.Labels) error {
+	//load corresponding repo, account and manifest
+	repo, err := keppel.FindRepositoryByID(tx, securityInfo.RepositoryID)
+	if err != nil {
+		return fmt.Errorf("cannot find repo for manifest %s: %s", securityInfo.Digest, err.Error())
+	}
+	account, err := keppel.FindAccount(tx, repo.AccountName)
+	if err != nil {
+		return fmt.Errorf("cannot find account for repo %s: %s", repo.FullName(), err.Error())
+	}
+	manifest, err := keppel.FindManifest(tx, *repo, securityInfo.Digest)
+	if err != nil {
+		return fmt.Errorf("cannot find manifest for repo %s and digest %s: %s", repo.FullName(), securityInfo.Digest, err.Error())
+	}
+
+	err = j.doSecurityCheck(*account, *repo, *manifest, &securityInfo)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Update(&securityInfo)
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func (j *Janitor) doSecurityCheck(account keppel.Account, repo keppel.Repository, manifest keppel.Manifest, securityInfo *keppel.TrivySecurityInfo) (returnedError error) {
+	//clear timing information (this will be filled down below once we actually talk to Trivy;
+	//if any preflight check fails, the fields stay at nil)
+	securityInfo.CheckedAt = nil
+	securityInfo.CheckDurationSecs = nil
+
+	//skip validation while account is in maintenance (maintenance mode blocks
+	//all kinds of activity on an account's contents)
+	if account.InMaintenance {
+		securityInfo.NextCheckAt = j.timeNow().Add(j.addJitter(1 * time.Hour))
+		return nil
+	}
+
+	continueCheck, err := j.checkPreConditionsForTrivy(account, repo, manifest, securityInfo)
+	if err != nil {
+		return err
+	}
+	if !continueCheck {
+		return nil
+	}
+
+	//we know that this image will not be "Unsupported", so the rest is the part where we actually
+	//talk to Trivy (well, mostly anyway), so that part deserves to be measured for performance
+	checkStartedAt := j.timeNow()
+	defer func() {
+		if returnedError == nil {
+			checkFinishedAt := j.timeNow()
+			securityInfo.CheckedAt = &checkFinishedAt
+			duration := checkFinishedAt.Sub(checkStartedAt).Seconds()
+			securityInfo.CheckDurationSecs = &duration
+		}
+	}()
+
+	imageRef := keppel.ImageReference{
+		Host:      j.cfg.APIPublicHostname,
+		RepoName:  fmt.Sprintf("%s/%s", account.Name, repo.Name),
+		Reference: keppel.ManifestReference{Digest: manifest.Digest},
+	}
+
+	tokenResp, err := auth.Authorization{
+		UserIdentity: auth.AnonymousUserIdentity,
+		Audience:     auth.Audience{},
+		ScopeSet: auth.NewScopeSet(auth.Scope{
+			ResourceType: "repository",
+			ResourceName: repo.FullName(),
+			Actions:      []string{"pull"},
+		}),
+	}.IssueTokenWithExpires(j.cfg, 20*time.Minute)
+	if err != nil {
+		return err
+	}
+
+	//ask Trivy for the security status of the manifest
+	securityInfo.Message = "" //unless it gets set to something else below
```

Comment:
The Message field does not get populated on errors. Tying into my previous review comment, I suggest filling Message on error in the deferred function:

```go
if err != nil {
  securityInfo.Status = clair.ErrorVulnerabilityStatus
  securityInfo.Message = err.Error()
  securityInfo.NextCheckAt = j.timeNow().Add(j.addJitter(5 * time.Minute))
}
```